taichi/program/state_flow_graph.cpp

```diff
@@ -248,17 +248,21 @@ bool StateFlowGraph::optimize_listgen() {
         // Test if two list generations share the same mask and parent list
         auto snode = node_a->meta->snode;
 
-        auto mask_state = AsyncState{snode, AsyncState::Type::mask};
         auto list_state = AsyncState{snode, AsyncState::Type::list};
         auto parent_list_state =
             AsyncState{snode->parent, AsyncState::Type::list};
 
-        TI_ASSERT(get_or_insert(node_a->input_edges, mask_state).size() == 1);
-        TI_ASSERT(get_or_insert(node_b->input_edges, mask_state).size() == 1);
+        if (!snode->is_path_all_dense) {
+          // Needs mask state
+          auto mask_state = AsyncState{snode->get_least_sparse_ancestor(),
+                                       AsyncState::Type::mask};
+          TI_ASSERT(get_or_insert(node_a->input_edges, mask_state).size() == 1);
+          TI_ASSERT(get_or_insert(node_b->input_edges, mask_state).size() == 1);
```

Comment:
If we have `ti.pointer.dense.dense.dense`, it seems to me that the `dense` nodes should not need mask states (of their `get_least_sparse_ancestor()`) for list generation. No sparsity tests will happen anyway for list generation of dense SNodes. The list state of their parent already defines which child elements should be inserted into their list.

This also depends on how we create `TaskMeta` for `listgen` tasks - it seems to me that `snode->is_path_all_dense` should not be the condition for using mask state or not, instead we should simply test if the SNode itself is dense or sparse.